[Inductor max autotune] Multithreaded Precompilation

kadeng opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #119010
* #119009
* #119008
* #119007
* #119006
* #119004
* __->__ #119003

When using the Cutlass backend, the compilation
of CUDA source files can totally dominate the runtime required for the benchmarking done
as part of Autotuning.

This change adds a multithreaded precompilation phase, which serves to pre-populate the compilation cache ( both in-memory, and a
possible on-disk sccache ).

Also it ensures that no unneccessary compilation
and benchmarking steps are performed, which was peviously the case.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @muchulee8 @aakhundov @ColinPeppler

kadeng committed:
[Inductor max autotune] Multithreaded Precompilation

When using the Cutlass backend, the compilation
of CUDA source files can totally dominate the runtime required for the benchmarking done
as part of Autotuning.

This change adds a multithreaded precompilation phase, which serves to pre-populate the compilation cache ( both in-memory, and a
possible on-disk sccache ).

Also it ensures that no unneccessary compilation
and benchmarking steps are performed, which was peviously the case.

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor max autotune] Multithreaded Precompilation"


When using the Cutlass backend, the compilation
of CUDA source files can totally dominate the runtime required for the benchmarking done
as part of Autotuning.

This change adds a multithreaded precompilation phase, which serves to pre-populate the compilation cache ( both in-memory, and a
possible on-disk sccache ).

Also it ensures that no unneccessary compilation
and benchmarking steps are performed, which was peviously the case.

[ghstack-poisoned]

jansel commented on torch/_inductor/codecache.py:
```diff
@@ -292,8 +293,18 @@ def check_cache(cache, callback=None) -> bool:
                 and check_cache(self.get_global_cache(), callback=log_stats)
             ):
                 try:
-                    # re-benchmark everything to try to get consistent numbers from the same machine
```
This comment seems important.  I think we do want consistent numbers from the same machine.  If you mix old numbers with new numbers it may not be apples-to-apples.

Typically, compiling is *much* more expensive than running.  So the cost of re-running (with a cached compile) should be very low.

kadeng replied:
I will change that back if you like. It will literally mean that we run the same code again on the same machine, just at a different point in time. So I can imagine there is a difference if GPU / CPU power settings or background load have changed in the meantime, but not really much else.

kadeng replied:
done

jansel commented on torch/_inductor/config.py:
```diff
@@ -200,6 +200,18 @@ def is_fbcode():
+    int(
+        os.environ.get(
+            "TORCHINDUCTOR_MAX_AUTOTUNE_PRECOMPILATION_WORKERS",
+            str(torch.get_num_threads() - 8),
```
Why -8?  This could be negative on some machines.

kadeng replied:
Yes, negative values are allowed, in that case we simply don't have parallel precompilation. The idea is to leave at least 8 CPUs unused by parallel precompilation by default. But we could change that of course.

jansel replied:
Why not use just use torch.get_num_threads()?

kadeng replied:
The idea is to leave 8 CPUs unused under the assumption that these are required for whatever else is happening on the system. But I will remove the -8.

jansel commented on torch/_inductor/select_algorithm.py:
```diff
@@ -765,7 +767,55 @@ def __call__(
+                len(choices),
+                num_workers,
+            )
+            with ThreadPoolExecutor(max_workers=num_workers) as executor:
```
Why not reuse the same threadpool and infra for parallel compiles?  This seems like it is doing the same thing as async compile...

kadeng replied:
Threads are relatively lightweight to construct but costly to keep alive, so there's no real benefit but a downside to maintain a persistent threadpool. This code works well and is compact and doesn't couple otherwise unrelated code sections design-wise. AsyncCompile makes much more sense for Trito. Triton requires python processes to compile the code which have the GIL issue. For these processes, persistence makes much more sense. The threadpool in AsyncCompile mostly makes sense because it allows to wait on **both** the futures from the Triton compiler processes and the CUDA compilation threads.

jansel reviewed (changes requested):
Add tests

kadeng commented:
> Add tests



> Add tests

This is actually on the default codepath for all unit tests that involve max autotune. So while there are not dedicated *unit* tests of the precompilation functionality per-se, a lot of tests would immediately fail if this would not work. Performance benefits in practice are proven through manual tests. But I will add a test where I force the config setting..

kadeng committed:
Update on "[Inductor max autotune] Multithreaded Precompilation"


When using the Cutlass backend, the compilation
of CUDA source files can totally dominate the runtime required for the benchmarking done
as part of Autotuning.

This change adds a multithreaded precompilation phase, which serves to pre-populate the compilation cache ( both in-memory, and a
possible on-disk sccache ).

Also it ensures that no unneccessary compilation
and benchmarking steps are performed, which was peviously the case.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor max autotune] Multithreaded Precompilation"


When using the Cutlass backend, the compilation
of CUDA source files can totally dominate the runtime required for the benchmarking done
as part of Autotuning.

This change adds a multithreaded precompilation phase, which serves to pre-populate the compilation cache ( both in-memory, and a
possible on-disk sccache ).

Also it ensures that no unneccessary compilation
and benchmarking steps are performed, which was peviously the case.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor max autotune] Multithreaded Precompilation"


When using the Cutlass backend, the compilation
of CUDA source files can totally dominate the runtime required for the benchmarking done
as part of Autotuning.

This change adds a multithreaded precompilation phase, which serves to pre-populate the compilation cache ( both in-memory, and a
possible on-disk sccache ).

Also it ensures that no unneccessary compilation
and benchmarking steps are performed, which was peviously the case.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor max autotune] Multithreaded Precompilation"


When using the Cutlass backend, the compilation
of CUDA source files can totally dominate the runtime required for the benchmarking done
as part of Autotuning.

This change adds a multithreaded precompilation phase, which serves to pre-populate the compilation cache ( both in-memory, and a
possible on-disk sccache ).

Also it ensures that no unneccessary compilation
and benchmarking steps are performed, which was peviously the case.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng commented:
@pytorchbot merge

kadeng commented:
Accidentally clicked merge, which doesn't work with ghstack. I can't reopen it, so I had to recreate the PR here: https://github.com/pytorch/pytorch/pull/119386
